ostd/src/cpu/cpu_local.rs

```diff
@@ -0,0 +1,223 @@
+//! being not [`Copy`], have a constant constructor [`Option::None`] that
+//! produces a value that can be bitwise copied to create a new instance.
+//! [`alloc::sync::Arc`] however, don't have such a constructor, and thus cannot
+//! be directly used as a CPU-local object. Wrapping it in a type that has a
+//! constant constructor, like [`Option<T>`], can make it CPU-local.
+
+use core::ops::Deref;
+
+use crate::{
+    cpu::{get_cpu_local_base, set_cpu_local_base},
+    trap::{disable_local, DisabledLocalIrqGuard},
+};
+
+/// Defines a CPU-local variable.
+///
+/// # Example
+///
+/// ```rust
+/// use crate::cpu_local;
+/// use core::cell::RefCell;
+///
+/// cpu_local! {
+///     static FOO: RefCell<u32> = RefCell::new(1);
+///
+///     #[allow(unused)]
+///     pub static BAR: RefCell<f32> = RefCell::new(1.0);
+/// }
+///
+/// println!("FOO VAL: {:?}", *FOO.borrow());
+/// ```
+#[macro_export]
+macro_rules! cpu_local {
+    ($( $(#[$attr:meta])* $vis:vis static $name:ident: $t:ty = $init:expr; )*) => {
+        $(
+            #[link_section = ".cpu_local"]
+            $(#[$attr])* $vis static $name: $crate::CpuLocal<$t> = $crate::CpuLocal::__new($init);
+        )*
+    };
+}
+
+extern "C" {
+    fn __cpu_local_start();
+    fn __cpu_local_end();
+}
+
+/// CPU-local objects.
+///
+/// A CPU-local object only gives you immutable references to the underlying value.
+/// To mutate the value, one can use atomic values (e.g., [`AtomicU32`]) or internally mutable
+/// objects (e.g., [`RefCell`]).
+///
+/// [`AtomicU32`]: core::sync::atomic::AtomicU32
+/// [`RefCell`]: core::cell::RefCell
+pub struct CpuLocal<T>(T);
+
+// SAFETY: At any given time, only one task can access the inner value T
+// of a cpu-local variable even if `T` is not `Sync`.
+unsafe impl<T> Sync for CpuLocal<T> {}
+
+// Prevent valid instances of CpuLocal from being copied to any memory
+// area outside the .cpu_local section.
+impl<T> !Copy for CpuLocal<T> {}
+impl<T> !Clone for CpuLocal<T> {}
+
+// In general, it does not make any sense to send instances of CpuLocal to
+// other tasks as they should live on other CPUs to make sending useful.
+impl<T> !Send for CpuLocal<T> {}
+
+impl<T> CpuLocal<T> {
+    /// Initialize a CPU-local object.
+    ///
+    /// Please do not call this function directly. Instead, use the
+    /// `cpu_local!` macro.
+    ///
+    /// # Safety
+    ///
+    /// The caller should ensure that the object initialized by this
+    /// function resides in the `.cpu_local` section. Otherwise the
+    /// behavior is undefined.
+    ///
+    /// The reason why whe don't mark it unsafe is that instantiating
+    /// a `CpuLocal` object with the macro is safe, but wrapping it
+    /// in an unsafe block would allow you to write unsafe code when
+    /// constructing `val`. So, again, please don't call this function
+    /// directly.
+    #[doc(hidden)]
+    pub const fn __new(val: T) -> Self {
```

Comment:
Thanks for writing up the rationale regarding why the `unsafe` keyword is removed from the function defitition, although the rationale itself is well... insufficient.

The constructor method must be unsafe and the macro can be fixed.

```rust
macro_rules! cpu_local {
    ($( $(#[$attr:meta])* $vis:vis static $name:ident: $t:ty = $init:expr; )*) => {
        $(
            #[link_section = ".cpu_local"]
            $(#[$attr])* $vis static $name: $crate::CpuLocal<$t> = {
                let val = $init;
                // SAFETY: The CPU local variable is statically stored in
                // the special .cpu_local section.
                unsafe {
                    $crate::CpuLocal::__new(val)
                }
            }
        )*
    };
}
```

Reply:
wow